soccer/planning/SettlePathPlanner.cpp

```diff
@@ -0,0 +1,419 @@
+    if (averageBallVel.angleBetween(ball.vel) > maxBallAngleChangeForPathReset ||
+        (averageBallVel - ball.vel).mag() > *_maxBallVelForPathReset) {
+        firstTargetPointFound = false;
+        firstBallVelFound = false;
+    }
+
+
+    // Smooth out the ball velocity a little bit so we can get a better estimate of intersect points
+    if (firstBallVelFound) {
+        averageBallVel = *_ballVelGain * ball.vel +
+                         (1 - *_ballVelGain) * averageBallVel;
+    } else {
+        averageBallVel = ball.vel;
+        firstBallVelFound = true;
+    }
+
+    // Change start instant to be the partial path end instead of the robot current location
+    // if we actually have already calculated a path the frame before
+    if (prevPath) {
+        timeIntoPreviousPath = curTime - prevPath->startTime();
+
+        // Make sure we still have time in the path to replan and correct
+        // since it's likely that the old path is slightly off
+        //
+        // ---|----------------|-----------------|
+        // TimeNow     EndPartialPrevPath  FinalTargetPoint
+        //                     |-----------------|
+        //          Amount of the path we can change this iteration
+        if (timeIntoPreviousPath < prevPath->getDuration() - partialReplanLeadTime &&
+            timeIntoPreviousPath > 0ms) {
+            partialPath =
+                prevPath->subPath(0ms, timeIntoPreviousPath + partialReplanLeadTime);
+            partialPathTime = partialPath->getDuration() - timeIntoPreviousPath;
+            startInstant = partialPath->end().motion;
+        }
+    }
+
+    // State transitions
+    // Intercept -> Dampen, PrevPath and almost at the end of the path
+    // Dampen -> Complete, PrevPath and almost slowed down to 0?
+    if (prevPath && timeIntoPreviousPath < prevPath->getDuration()) {
+        Geometry2d::Line ballMovementLine(ball.pos, ball.pos + averageBallVel);
+        std::unique_ptr<Path> pathSoFar = prevPath->subPath(0ms, timeIntoPreviousPath);
+        
+        float botDistToBallMovementLine = ballMovementLine.distTo(pathSoFar->end().motion.pos);
+
+        // Intercept -> Dampen
+        //  Almost interescting the ball path
+        //    or Almost at end of the target path
+        //  Actually in front of the ball
+        //
+        // TODO: Check ball sense?
+        if (timeIntoPreviousPath < prevPath->getDuration() - 1.5*partialReplanLeadTime &&
+            (pathSoFar->end().motion.pos - prevPath->end().motion.pos).mag() < 3*Robot_Radius &&
+            (pathSoFar->end().motion.pos - ballMovementLine.pt[0]).mag() >
+                (pathSoFar->end().motion.pos - ballMovementLine.pt[1]).mag() &&
+            currentState == Intercept) {
+            
+            startInstant = pathSoFar->end().motion;
+            currentState = Dampen;
+            std::cout << "Transitioned to dampen" << std::endl;
+        }
+    }
+
+    switch (currentState) {
+    case Intercept: {
```

Comment:
Similar comment about switch-cases here. These seem a bit complex to be nested in an already large switch statement, maybe refactor to helper functions? (something something cyclomatic complexity?)